### PR TITLE
AlertingNG: Add Telegram notification channel

### PR DESCRIPTION
**Pull request by codesome**
**What this PR does / why we need it**:

Adds Telegram notification channel. Many things are taken from Grafana implementation.

~Once #32604 is merged, I will rebase this PR to plug Telegram into the notification pipeline~

### CODE REVIEW

File: pkg/services/ngalert/notifier/channels/telegram.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	ChatID      string
+	Message     string
+	log         log.Logger
+	tmpl        *template.Template
+	externalUrl *url.URL
+}
+
+// NewTelegramNotifier is the constructor for the Telegram notifier
+func NewTelegramNotifier(model *models.AlertNotification, t *template.Template, externalUrl *url.URL) (*TelegramNotifier, error) {
+	if model.Settings == nil {
```
Comment (gotjosh) [inline pkg/services/ngalert/notifier/channels/telegram.go:40]: This is the first time I see this validation, should we also add it to the others?

Comment (codesome) [reply pkg/services/ngalert/notifier/channels/telegram.go:40]: I am basically copy pasting existing code and making changes. So looks like we will need this for others too which is missing currently. I will add in the other PRs and update this PR for the other notifiers.

File: pkg/services/ngalert/notifier/channels/telegram.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	chatID := model.Settings.Get("chatid").MustString()
+	message := model.Settings.Get("message").MustString(`
+{{ if gt (len .Alerts.Firing) 0 }}
+**Firing**
+{{ template "__text_alert_list" .Alerts.Firing }}
+
+{{ end }}
+{{ if gt (len .Alerts.Resolved) 0 }}
+**Resolved**
+{{ template "__text_alert_list" .Alerts.Resolved }}
```
[the remark this answers is not in the document]
Comment (codesome) [reply pkg/services/ngalert/notifier/channels/telegram.go:54]: `__text_alert_list` is already present in the file above, hence you dont see any addition here

File: pkg/services/ngalert/notifier/channels/telegram.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		return err
+	}
+	if _, err := fw.Write([]byte(value)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (tn *TelegramNotifier) SendResolved() bool {
+	return !tn.GetDisableResolveMessage()
```
[the remark this answers is not in the document]
Comment (codesome) [reply pkg/services/ngalert/notifier/channels/telegram.go:155]: There is a preexisting `NotifierBase` which is embedded in the struct (same as current notifiers in 7.x) which has all the default setting. We are initialising it in the New() function. So this method is coming from there and coming from setting.

File: pkg/services/ngalert/notifier/channels/telegram.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	msg, err := tn.buildTelegramMessage(ctx, as)
+	if err != nil {
+		return false, err
+	}
+
+	var body bytes.Buffer
+	w := multipart.NewWriter(&body)
+	wClosed := false
+	defer func() {
+		if !wClosed {
```
[the remark this answers is not in the document]
Comment (codesome) [reply pkg/services/ngalert/notifier/channels/telegram.go]: > Under which scenario would this happen

If the writeField fails below.

So we want to close it before we use it. Closing the writer adds the last boundary to multipart writer. So we cannot have it in the defer alone. But I see `wClosed` is not require, double close should be fine.

**Review by gotjosh — APPROVED** [review]

**Pull request merged**